components/common/block_throttle/component.adb

```diff
@@ -0,0 +1,339 @@
+                                                          Re);
+                           case Re is
+                              when Block.Success =>
+                                 Block_Client.Enqueue (Client, Cache (I).C);
+                              when Block.Retry =>
+                                 null;
+                              when others =>
+                                 Cache (I).A := True;
+                           end case;
+                        when Block.Allocated =>
+                           if Block_Client.Assigned (Client, Cache (I).C) then
+                              Block_Client.Enqueue (Client, Cache (I).C);
+                           end if;
+                        when Block.Pending =>
+                           if Block_Client.Assigned (Client, Cache (I).C) then
+                              Block_Client.Update_Request (Client, Cache (I).C);
+                              Pr := Pr or else Block_Client.Status (Cache (I).C) /= Block.Pending;
+                           end if;
+                        when Block.Ok | Block.Error =>
+                           if
+                              Block_Client.Assigned (Client, Cache (I).C)
+                              and then not Cache (I).Q
+                           then
+                              if
+                                 Block_Client.Status (Cache (I).C) = Block.Ok
+                                 and then Block_Client.Kind (Cache (I).C) = Block.Read
+                              then
+                                 Block_Client.Read (Client, Cache (I).C);
+                              end if;
+                              if Current < Rate and then Block_Server.Assigned (Server, Cache (I).S) then
+                                 Block_Server.Acknowledge (Server, Cache (I).S, Block_Client.Status (Cache (I).C));
+                                 if Block_Server.Status (Cache (I).S) = Block.Raw then
+                                    Current := Current + 1;
+                                    Pr := True;
+                                 end if;
+                              else
+                                 Queue.Put (Ack_Queue, I);
+                                 Cache (I).Q := True;
+                                 Pr := True;
+                              end if;
+                           end if;
+                     end case;
+                  when Block.Error =>
+                     if Block_Server.Assigned (Server, Cache (I).S) then
+                        Block_Server.Acknowledge (Server, Cache (I).S, Block.Error);
+                        Pr := Pr or else Block_Server.Status (Cache (I).S) = Block.Raw;
+                     end if;
+                  when others =>
+                     null;
+               end case;
+            end loop;
+            Block_Client.Submit (Client);
+         end loop;
+         Block_Server.Unblock_Client (Server);
+      end if;
+   end Event;
```

Comment:
This subprogram is rather long. I wonder whether a different pattern can be used to implement this? Maybe handling the cases of `case Block_Server.Status (Cache (I).S) is` in separate subprograms?

Reply:
NB: This is more a general question. While it may still be acceptable in this case, I wonder how we'd structure more complex components (plus, those small components may serve as boilerplate for others).

Reply:
Fixed in https://github.com/Componolit/gravel/pull/6/commits/ed50d59d2d47bc1247dd653ffd17362a86aad39b.

Reply:
The structure of the component initialization will be refined in https://github.com/Componolit/gneiss/issues/114.